dll/win32/shell32/CDefView.cpp

```diff
@@ -767,40 +833,239 @@ void CDefView::UpdateListColors()
+    ERR("Unable to determine folder column from list column %d\n", (int) ListCol);
+    return hr;
+}
+
+HRESULT CDefView::LoadColumn(UINT FoldCol, UINT ListCol, BOOL Insert)
+{
+    WCHAR buf[MAX_PATH];
+    SHELLDETAILS sd;
+    HRESULT hr;
+
+    sd.str.uType = !STRRET_WSTR; // Protect ourself from broken implementations
+    hr = GetDetailsByFolderColumn(NULL, FoldCol, sd);
+    if (FAILED(hr))
+        return hr;
+    hr = StrRetToStrNW(buf, _countof(buf), &sd.str, NULL);
+    if (FAILED(hr))
+        return hr;
+
+    LVCOLUMN lvc;
+    lvc.mask = LVCF_TEXT | LVCF_FMT | LVCF_WIDTH | LVCF_SUBITEM;
+    lvc.pszText = buf;
+    lvc.fmt = sd.fmt;
+    lvc.cx = sd.cxChar * 8; // FIXME: Font size? DPI?
+    lvc.iSubItem = FoldCol; // Used by MapFolderColumnToListColumn & MapListColumnToFolderColumn
+    if ((int) ListCol == -1)
+    {
+        assert(Insert); // You can insert at the end but you can't change something that is not there
+        if (Insert)
+            ListCol = 0x7fffffff;
+    }
+    if (Insert)
+        ListView_InsertColumn(m_ListView.m_hWnd, ListCol, &lvc);
+    else
+        ListView_SetColumn(m_ListView.m_hWnd, ListCol, &lvc);
+    return S_OK;
+}
+
+HRESULT CDefView::LoadColumns(UINT *pColList, UINT ColListCount)
+{
+    HWND hWndHdr = ListView_GetHeader(m_ListView.m_hWnd);
+    UINT newColCount = 0, oldColCount = Header_GetItemCount(hWndHdr);
+    UINT foldCol, i;
+    HRESULT hr = S_FALSE;
+
+    m_ListView.SetRedraw(FALSE);
+    for (i = 0, foldCol = 0;; ++foldCol)
+    {
+        if (newColCount >= 0xffff)
+            break; // CompareIDs limit reached
+
+        if (pColList)
         {
-            if (FAILED(m_pSF2Parent->GetDetailsOf(NULL, i, &sd)))
+            if (i >= ColListCount)
                 break;
-            StrRetToStrNW( szTemp, 50, &sd.str, NULL);
-            m_ListView.InsertColumn(i, szTemp, sd.fmt, sd.cxChar * 8);
+            foldCol = pColList[i++];
+        }
 
-            InsertMenuW(m_hMenuArrangeModes, -1, MF_STRING, 0x30 + i,  szTemp);
+        SHCOLSTATEF state = 0;
+        if (!m_pSF2Parent || FAILED(m_pSF2Parent->GetDefaultColumnState(foldCol, &state)))
+            state = SHCOLSTATE_TYPE_STR | SHCOLSTATE_ONBYDEFAULT;
+
+        if (foldCol == 0)
+        {
+            // Force the first column
+        }
+        else if (state & SHCOLSTATE_HIDDEN)
+        {
+            continue;
+        }
+        else if (!pColList && !(state & SHCOLSTATE_ONBYDEFAULT))
+        {
+            continue;
         }
 
-        InsertMenuW(m_hMenuArrangeModes, -1, MF_BYPOSITION | MF_SEPARATOR, 0, 0);
+        bool insert = newColCount >= oldColCount;
+        UINT listCol = insert ? -1 : newColCount;
+        hr = LoadColumn(foldCol, listCol, insert);
+        if (FAILED(hr))
+        {
+            if (!pColList)
+                hr = S_OK; // No more items, we are done
+            break;
+        }
+        ++newColCount;
     }
-    else
+    for (i = newColCount; i < oldColCount; ++i)
```

Comment:
Will this ever be used? So far this function is called only on initialization and as far as I understand there are no columns in the list at this point.

Reply:
When the More dialog specifies fewer columns than what the list currently has.